### PR TITLE
update default args+kwargs hashing

### LINKED ISSUE #136: workaround for `TypeError: unhashable type: 'list'`
Hello, I think that `list` as an argument is quite often, so how about hashing and converting it internally for hashing only to `tuple`, where the order is preserved... what do you think about adding it to the codebase?

### PR DESCRIPTION
**Pull request by Borda**
resolves #136

### CODE REVIEW

**Review by shaypal5 — CHANGES_REQUESTED** [review]
I don't know. List and tuples are different Python objects.

What if somebody has a function where both lists and tuples are valid arguments to one of the parameters, but their interpretations are different? We practically erase some of the functionality!

This should be something you can opt into, using an environment variable or a one-time global function call, not automatic.

Please amend accordingly.

**Comment by Borda** [comment]
I may see what you mean, but not sure, so would it be better to move it to the default function so no effect for users who already have their own hashing?
also, the hash may reflect the initial data type; what do you think?

**Comment by shaypal5** [comment]
Regarding your first suggestion: No, that's not enough.

Regarding your second on: Ok, that interesting. If you append each tuple-from-list with a custom marker object, it will always have a different hash.

I mean that real tuples get hashed as: `(1, 'b')`.

But the a list of `[1, 'b']` should be converted so:
```python
class ListTupleMarker:
    pass

LIST_TUPLE_MARKER = ListTupleMarker()

# internally this is (1, 'b', LIST_TUPLE_MARKER)
list_as_tuple = (*as_list, LIST_TUPLE_MARKER)
```

**Comment by Borda** [comment]
The method that uses pickling to serialize the arguments will indeed preserve argument types because pickling captures the data structure along with its type information. When you pickle a Python object, the resulting byte stream fully encodes the object's structure and the types of its contents.

Therefore, when you serialize args and kwargs using pickle.dumps(), each serialized form represents the combination of values and their types. When you subsequently generate a hash from this serialized form, the hash is implicitly sensitive not only to the values but also to their types.

Here's an example demonstrating this:

```python
import pickle
import hashlib

# Integer and string with same "value"
arg_int = 1
arg_str = '1'

# Serialize each argument
serialized_int = pickle.dumps(arg_int)
serialized_str = pickle.dumps(arg_str)

# They produce different byte streams
assert serialized_int != serialized_str

# So their hashes will also be different
hash_int = hashlib.sha256(serialized_int).hexdigest()
hash_str = hashlib.sha256(serialized_str).hexdigest()

assert hash_int != hash_str

print(f'Hash for integer 1: {hash_int}')
print(f'Hash for string "1": {hash_str}')
```

File: cachier/core.py [hunk 1]
```diff
@@ -70,11 +69,12 @@ def _calc_entry(core, key, func, args, kwds):
 
 
 def _default_hash_func(args, kwds):
-    key = functools._make_key(args, kwds, typed=True)
```
Comment (Borda) [inline cachier/core.py:73]: so it also does not rely on private API, which can change any time

**Comment by Borda** [comment]
@shaypal5, mind check it now? :flags:

**Review by shaypal5 — APPROVED** [review]
Nice work.

**Pull request merged**